Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -154,31 +161,108 @@ namespace ROS2
         message.header.frame_id = ros2Frame->GetFrameID().data();
         message.header.stamp = ROS2Interface::Get()->GetROSTimestamp();
         message.height = 1;
-        message.width = lastScanResults.m_points.size();
-        message.point_step = sizeof(AZ::Vector3);
-        message.row_step = message.width * message.point_step;
+        message.width = pointCount;
+        message.point_step = sizeof(PackedVector3);
 
-        AZStd::array<const char*, 3> point_field_names = { "x", "y", "z" };
-        for (int i = 0; i < point_field_names.size(); i++)
-        {
+        AZStd::array<const char *, 3> pointFieldNames = {"x", "y", "z"};
+
+        for (int i = 0; i < pointFieldNames.size(); i++) {
             sensor_msgs::msg::PointField pf;
-            pf.name = point_field_names[i];
+            pf.name = pointFieldNames[i];
             pf.offset = i * 4;
             pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
             pf.count = 1;
             message.fields.push_back(pf);
         }
 
-        size_t sizeInBytes = lastScanResults.m_points.size() * sizeof(AZ::Vector3);
+        if (lastScanResults.m_ids.has_value()) {
+            sensor_msgs::msg::PointField pfId;
+            pfId.name = "entity_id";
+            pfId.offset = message.point_step;
+            pfId.datatype = sensor_msgs::msg::PointField::INT32;
+            pfId.count = 1;
+            message.fields.push_back(pfId);
+            constexpr auto fieldLength = sizeof(int32_t);
+            message.point_step += fieldLength;
+        }
+        AZStd::array<AZ::Color, 256> colorLookupTable;
+        if (lastScanResults.m_classes.has_value()) {
+            colorLookupTable = m_lidarCore.m_lidarConfiguration.
+                    GenerateSegmentationColorsLookupTable();
+
+            AZStd::array<const char *, 3> colorFieldNames = {"r", "g", "b"};
+
+            for (int i = 0; i < colorFieldNames.size(); i++) {
+                sensor_msgs::msg::PointField pf;
+                pf.name = colorFieldNames[i];
+                pf.offset = message.point_step;
+                pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
+                pf.count = 1;
+                message.fields.push_back(pf);
+                message.point_step += sizeof(float);
+            }
+
+            sensor_msgs::msg::PointField pfClassId;
+            pfClassId.name = "class_id";
+            pfClassId.offset = message.point_step;
+            pfClassId.datatype = sensor_msgs::msg::PointField::UINT8;
+            pfClassId.count = 1;
+            message.fields.push_back(pfClassId);
+            message.point_step += sizeof(uint8_t);
+        }
+
+        const auto sizeInBytes = pointCount * message.point_step;
         message.data.resize(sizeInBytes);
+
+        PackedVector3 xyz = {0.0f, 0.0f, 0.0f};
+
+        // IMO this is the clearest way to handle the offset with future fields, compiler should optimize it
```

Comment:
Comment to remove after review. Optionally to rephrase, depending on what you would like to explain here.